### PR TITLE
Rewrite kerberos tests

### PR DESCRIPTION
**Pull request by hlinnaka**
This adds a whole new test script in src/test/regress, implemented in plain bash. It sets up temporary a KDC as part of the script, and doesn't therefore rely on a pre-existing Kerberos server, like the old MU_kerberos-smoke test job did. It does require MIT Kerberos server-side utilities to be installed, instead, but no server needs to be running, and no superuser privileges are required.

This supersedes the MU_kerberos-smoke behave tests. The new rewritten bash script tests the same things:
1. You cannot connect to the server before running 'kinit' (to verify that the server doesn't just let anyone in, which could happen if the pg_hba.conf is misconfigured for the test, for example)
2. You can connect, after running 'kinit'
3. You can no longer connect, if the user account is expired

The new test script is hooked up to the top-level installcheck-world target.

There were also some Kerberos-related tests in TINC. Remove all that, too. They didn't seem interesting in the first place, looks like they were just copies of a few random other tests, intended to be run as a smoke test, after a connection had been authenticated with Kerberos, but there was nothing in there to actually set up the Kerberos environment in TINC.

### CODE REVIEW

**Comment by hlinnaka** [comment]
Note: the test script runs "pg_ctl restart -w", which currently crashes because of an unrelated bug. See https://github.com/greenplum-db/gpdb/pull/1953. That PR needs to be pushed before this.

**Comment by hlinnaka** [comment]
There's also some Kerberos stuff in gpAux/extensions/gphdfs/regression/legacy/configs/kerberos/. I couldn't figure out if it's actually used for anything, or if it's dead. Is it related to setting up a test HDFS cluster, using Kerberos authentication, perhaps? Anyone know?

**Comment by ivannovick** [comment]
@hlinnaka VERY COOL!  I was involved in hacking the old version and like this much better

**Comment by larham** [comment]
Regarding the need to install the /usr/sbin/krb5kdc tool, the repo https://github.com/Pivotal-DataFabric/toolsmiths-images defines the pipeline's docker images, and has its own pipeline and PR system.  Inside that repo, file toolchain_common.bash has function install_kerberos() which seems like an appropriate place for adding an additional kerberos tool.

**Comment by Eulerizeit** [comment]
It would be good if all tests liked in the GPDB repo.  Getting stuff out of
other repos should be a top priority

Rob

On Mar 6, 2017 6:41 PM, "Larry Hamel" <notifications@github.com> wrote:

Regarding the need to install the /usr/sbin/krb5kdc tool, the repo
https://github.com/Pivotal-DataFabric/toolsmiths-images defines the
pipeline's docker images, and has its own pipeline and PR system. Inside
that repo, file toolchain_common.bash has function install_kerberos() which
seems like an appropriate place for adding an additional kerberos tool.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/greenplum-db/gpdb/pull/1954#issuecomment-284571544>,
or mute
the thread

**Comment by hlinnaka** [comment]
On 03/07/2017 01:41 AM, Larry Hamel wrote:
> Regarding the need to install the /usr/sbin/krb5kdc tool, the repo
> https://github.com/Pivotal-DataFabric/toolsmiths-images defines the
> pipeline's docker images, and has its own pipeline and PR system.
> Inside that repo, file toolchain_common.bash has function
> install_kerberos() which seems like an appropriate place for adding
> an additional kerberos tool.

Ok. Larry, could you help me drive this PR to completion, and hook it up 
to the concourse CI, to replace the current MU_kerberos_smoke job, 
please? And by "help me", I mean "do it for me" ;-)

- Heikki

**Comment by hlinnaka** [comment]
(sorry, didn't see this paragraph at first:)

On 03/07/2017 01:41 AM, Larry Hamel wrote:
> So now the question is: why does the script in this PR refer to
> /usr/sbin/krb5kdc instead of /usr/local/sbin/krb5kdc , or is this
> some kind of platform dependent path? Alternatively, on my sample
> container from today's existing pipeline, krb5kdc is on the path
> already.

Usually, if you install debian or rpm packages, krb5kdc will be in 
/usr/sbin. But if you install from sources, it'll go to /usr/local/sbin 
by default.

I guess we should make the test script work regardless of where it is. 
/usr/sbin nor /usr/local/sbin are not usually in $PATH for a 
non-superuser. Perhaps add both to $PATH, or have an extra environment 
variable that you can use to tell the script where it is.

- Heikki

**Comment by larham** [comment]
Yes, I (or whoever picks up the story in the M&M team's backlog) will create a test pipeline for this PR and experiment with $PATH options, with a goal of working on both possible paths.

File: src/test/regress/kerberos/setup_test.sh [hunk 1]
```diff
@@ -1,61 +0,0 @@
...
-TestConnectivity
-if [ -z "$CONNECTIVITY_RESULT" ]; then
-	echo "test setup_kerberos_kdestroy ... ok (0 sec)"
-else
-	echo "RESULT: $RESULT"
-	echo "test setup_kerberos_kdestroy ... FAILED (0 sec)"
-	exit 1
-fi
-
-LD_LIBRARY_PATH= kinit -k -t $HOME/gpdb-kerberos.keytab gpadmin/kerberos-test@KRB.GREENPLUM.COM
```
Comment (jimmyyih) [inline src/test/regress/kerberos/setup_test.sh]: So we added `LD_LIBRARY_PATH=` before kinit and kdestroy a couple years back due to some shared object dependency issue when having greenplum_path.sh sourced.  Not sure if that issue still exists or not...

Okay so just checked and it seems we still do:
```
$ kinit
kinit: relocation error: kinit: symbol krb5_get_init_creds_opt_set_anonymous, version krb5_3_MIT not defined in file libkrb5.so.3 with link time reference
```

Although this is mostly an issue with ancient artifacts obtained from artifactory to build Greenplum in the Concourse pipeline.  You should not see it in regular compilation using system libraries.  Preferably, I'd like to see artifactory be removed from the Concourse pipeline for Greenplum 5.0+.

**Comment by cjcjameson** [comment]
Hi all,

First off, I think we're fine to just use the PR pipeline, because this PR is moving the kerberos test into ICW. Your PR went red in the PR pipeline because of the `git describe` versioning change.

--> Can you rebase and fetch tags and push? Then the PR will get tested further in that pipeline.

We wanted to try to reproduce Heikki's success locally, since some of the above comments were about making it work both in CI and on local workstations.

we ran: `make -C src/test/kerberos/ installcheck`

and got back:
```
bash installcheck.sh
if [ -e ./krb5kdc.pid ]; then kill `cat ./krb5kdc.pid`; fi || true
rm -f ./krb5kdc.pid
rm -rf test-kdc-db
rm -f krb5cc server.keytab client.keytab
rm -f pg_hba.conf.orig pg_hba.conf.kerberized postgresql.conf.orig postgresql.conf.kerberized
Setting up test KDC...
Loading random data
Initializing database '/Users/pivotal/workspace/gpdb/src/test/kerberos/test-kdc-db/principal' for realm 'GPDB.EXAMPLE',
master key name 'K/M@GPDB.EXAMPLE'
/Users/pivotal/workspace/gpdb/src/test/kerberos
Authenticating as principal pivotal/admin@GPDB.EXAMPLE with password.
kadmin.local: Can not fetch master key (error: No such file or directory). while initializing kadmin.local interface
make: *** [installcheck] Error 1
```

We're wondering if there's some kerberos secrets, configuration, or initialization that we're missing. On stackoverflow, we saw some articles recommending that we create a new realm, but we weren't familiar with that and didn't see the binary tools that those posts referred to.

Our system: On a macOS, we installed kerberos 1.14 via brew. We did `brew link --force krb5`, partially as an accident, but it happened to make it easier to run the make targets without extra flags and environment variables. There is a chance that our environment has trouble because of interference between the built-in kerberos client and the brew-installed one.

**Comment by rvs** [comment]
@cjcjameson not to sound like a broken record, but using kerby http://directory.apache.org/kerby/ will make a lot of this pain go away. Even on the configuraiton side (secret management, etc.) kerby is much nicer than MIT native version. And, of course, it'll allow you to forget the difference of running a test on Mac OS vs. Linux vs... It just runs anywhere where Java is available.

**Comment by hlinnaka** [comment]
On 03/08/2017 04:04 AM, C.J. Jameson wrote:
> we ran: `make -C src/test/kerberos/ installcheck`
>
> and got back:
> ```
> bash installcheck.sh
> if [ -e ./krb5kdc.pid ]; then kill `cat ./krb5kdc.pid`; fi || true
> rm -f ./krb5kdc.pid
> rm -rf test-kdc-db
> rm -f krb5cc server.keytab client.keytab
> rm -f pg_hba.conf.orig pg_hba.conf.kerberized postgresql.conf.orig postgresql.conf.kerberized
> Setting up test KDC...
> Loading random data
> Initializing database '/Users/pivotal/workspace/gpdb/src/test/kerberos/test-kdc-db/principal' for realm 'GPDB.EXAMPLE',
> master key name 'K/M@GPDB.EXAMPLE'
> /Users/pivotal/workspace/gpdb/src/test/kerberos
> Authenticating as principal pivotal/admin@GPDB.EXAMPLE with password.
> kadmin.local: Can not fetch master key (error: No such file or directory). while initializing kadmin.local interface
> make: *** [installcheck] Error 1
> ```
>
> We're wondering if there's some kerberos secrets, configuration, or
> initialization that we're missing. On stackoverflow, we saw some
> articles recommending that we create a new realm, but we weren't
> familiar with that and didn't see the binary tools that those posts
> referred to.

Hmm. Everything should be included in the branch. I wonder if it's 
picking up some configuration files from /etc, that it shouldn't?

You can try running setup-kdc.sh directly, it's one step of 
installcheck.sh. In setup-kdc.sh, I think you got the error from this line:
```
/usr/sbin/kadmin.local -q "addprinc -randkey postgres/gpdb-server.example"
```

You could running that manually, with something like:
```
KRB5_KDC_PROFILE=./test-kdc-db/kdc.conf KRB5_CONFIG=./krb5.conf 
/usr/sbin/kadmin.local -q "addprinc -randkey postgres/gpdb-server.example"
```

The kerberos commands should create log files in test-kdc-db, see if 
there are any hints there.

You could also use strace to figure out what file it tries to open but 
fails.

> Our system: On a macOS, we installed kerberos 1.14 via brew. We did
> `brew link --force krb5`, partially as an accident, but it happened
> to make it easier to run the make targets without extra flags and
> environment variables. There is a chance that our environment has
> trouble because of interference between the built-in kerberos client
> and the brew-installed one.

That's a good thing to test. We want the script to be robust so that it 
works on whatever installation people happen to have on their 
development systems.

- Heikki

**Comment by cjcjameson** [comment]
Thanks everyone -- sorry for the radio silence -- we'll be getting back to this soon

**Comment by larham** [comment]
@hlinnaka fyi, a progress note: there's a branch at https://github.com/larham/gpdb/tree/rewrite-kerberos-tests with a dev pipeline (team=dev)  at  https://gpdb.ci.pivotalci.info/teams/dev/pipelines/gpdb_ace_rewrite-kerberos-tests.  This branch removes prefix paths for kerberos utilities , relying on them being in the $PATH, as they are for container's gpadmin user. It also restores kinit to have Jimmy's suggestion of a prefix `-LD_LIBRARY_PATH= kinit `... 

As you can see from the pipeline, the kerberos utilities give error:

**Comment by jimmyyih** [comment]
@larham, try adding `LD_LIBRARY_PATH= ` before kdb5_util.  Try it out manually to verify.

**Comment by larham** [comment]
Jimmy kindly suggested prefixing all kerberos utils with the LD_LIBRARY_PATH= prefix, which succeeded in getting past the "relocation" errors.

Currently, the test for a successful login is failing in the same way that the intentionally-unsuccessful login is failing:

```
Server configured for Kerberos authentication
Testing connection, should fail
psql: GSSAPI continuation error: An invalid name was supplied
GSSAPI continuation error: Hostname cannot be canonicalized
Obtaining Kerberos ticket-granting-ticket from KDC...
Testing connection, should succeed and print version
psql: GSSAPI continuation error: An invalid name was supplied
GSSAPI continuation error: Hostname cannot be canonicalized
make[1]: *** [installcheck] Error 2
make[1]: Leaving directory `/tmp/build/e18b2f02/gpdb_src/src/test/kerberos'
make: *** [installcheck-world] Error 2
```

**Comment by larham** [comment]
Hmm, and if the connection test fails, it seems like the kerberos server is left running, and the build pipeline thinks the job never finishes.

**Comment by larham** [comment]
We finally got green on a pipeline container, but it requires a change to DNS that has wide impact.

/etc/hosts needs an initial entry like:

```
# cat /etc/hosts
127.0.0.1 gpdb-server.example
127.0.0.1 localhost
10.254.0.10 62c2225e-46a3-45b5-6eba-1225be95a119
```

This implies that every pipeline container would have this entry, since we need to put it into a centralized script run by superuser (cannot just be done by the ICW script locally)

Further, for ICW to run on a dev box, each dev would need to have an entry like this on their workstation.

**Comment by hlinnaka** [comment]
On 03/22/2017 02:42 AM, Larry Hamel wrote:
> We finally got green on a pipeline container, but it requires a change to DNS that has wide impact.
>
> /etc/hosts needs an initial entry like:
>
> ```
> # cat /etc/hosts
> 127.0.0.1 gpdb-server.example
> 127.0.0.1 localhost
> 10.254.0.10 62c2225e-46a3-45b5-6eba-1225be95a119
> ```
>
> This implies that every pipeline container would have this entry, since we need to put it into a centralized script run by superuser (cannot just be done by the ICW script locally)
>
> Further, for ICW to run on a dev box, each dev would need to have an entry like this on their workstation.

That's odd, it worked fine on my system without those entries. What it 
is the error you get without those?

- Heikki

**Comment by larham** [comment]
with no entry, we get "canonicalized" errors
```
Testing connection, should fail
psql: GSSAPI continuation error: An invalid name was supplied
GSSAPI continuation error: Hostname cannot be canonicalized
Obtaining Kerberos ticket-granting-ticket from KDC...
Testing connection, should succeed and print version
psql: GSSAPI continuation error: An invalid name was supplied
GSSAPI continuation error: Hostname cannot be canonicalized
```

I believe this happens because 127.0.0.1 undergoes reverse DNS look-up to "localhost", which has no dot "." in it.

Putting an entry "127.0.0.1 gpdb-server.example" at the *top* of /etc/hosts provides this info.

When you ping your "localhost", what is the hostname? With the line above in /etc/hosts, we get a translation into "gpdb-server.example":

```
$ ping localhost
PING localhost (127.0.0.1) 56(84) bytes of data.
64 bytes from gpdb-server.example (127.0.0.1): icmp_seq=1 ttl=64 time=0.025 ms
```

**Comment by hlinnaka** [comment]
On 03/22/2017 05:56 PM, Larry Hamel wrote:
> with no entry, we get "canonicalized" errors
> ```
> Testing connection, should fail
> psql: GSSAPI continuation error: An invalid name was supplied
> GSSAPI continuation error: Hostname cannot be canonicalized
> Obtaining Kerberos ticket-granting-ticket from KDC...
> Testing connection, should succeed and print version
> psql: GSSAPI continuation error: An invalid name was supplied
> GSSAPI continuation error: Hostname cannot be canonicalized
> ```
>
> I believe this happens because 127.0.0.1 undergoes reverse DNS look-up to "localhost", which has no dot "." in it.
>
> Putting an entry "127.0.0.1 gpdb-server.example" at the *top* of /etc/hosts provides this info.

Hmm. A couple of things you could try to add to krb5.conf:

[libdefaults]
     rdns = false
     dns_canonicalize_hostname = false

[domain_realm]
     localhost = GPDB.EXAMPLE
     gpdb-server.example = GPDB.EXAMPLE


I'm not sure which ones of those might help, if any, so try adding them 
one by one and see what works..

> When you ping your "localhost", what is the hostname? With the line above in /etc/hosts, we get a translation into "gpdb-server.example":
>
> ```
> $ ping localhost
> PING localhost (127.0.0.1) 56(84) bytes of data.
> 64 bytes from gpdb-server.example (127.0.0.1): icmp_seq=1 ttl=64 time=0.025 ms
> ```

$ ping localhost
PING localhost(localhost (::1)) 56 data bytes
64 bytes from localhost (::1): icmp_seq=1 ttl=64 time=0.034 ms
64 bytes from localhost (::1): icmp_seq=2 ttl=64 time=0.054 ms
^C

- Heikki

**Comment by Chibin** [comment]
We took another stab at it by using `localhost` instead of `gpdb-server.example` and it looks like it's going through. Are there any concerns into changing it to `localhost`? If FQDN is not needed, then I feel like this should be fine as people would no longer need to modify their system just to make the kerberos test work.

**Comment by hlinnaka** [comment]
On 22 March 2017 20:13:53 EET, Marbin Tan <notifications@github.com> wrote:
>We took another stab at it by using `localhost` instead of
>`gpdb-server.example` and it looks like it's going through. Are there
>any concerns into changing it to `localhost`? If FQDN is not needed,
>then I feel like this should be fine as people would no longer need to
>modify their system just to make the kerberos test work.

Ok, works for me. After all, we're not trying to test Kerberos itself, just the integration with the libraries.

- Heikki

**Comment by Chibin** [comment]
@hlinnaka Is it okay to close this PR and create a new PR with the new changes that we've applied? or would you prefer to grab our changes from our branch?

**Comment by hlinnaka** [comment]
Sure.

**Comment by Chibin** [comment]
> Sure.

Thanks. Closing this out and created a new PR https://github.com/greenplum-db/gpdb/pull/2087

**Pull request closed without merging**